src/NUnitFramework/framework/Constraints/AnyOfConstraint.cs

```diff
@@ -0,0 +1,172 @@
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace NUnit.Framework.Constraints
+{
+    /// <summary>
+    /// <see cref="AnyOfConstraint"/> is used to determine whether item is equal to any of parameters
+    /// </summary>
+    public class AnyOfConstraint : Constraint
+    {
+        private readonly IEnumerable _expected;
+
+        /// <summary>
+        /// The NUnitEqualityComparer in use for this constraint
+        /// </summary>
+        private readonly NUnitEqualityComparer _comparer = NUnitEqualityComparer.Default;
+
+        /// <summary>
+        /// Construct a OneOfConstraint
+        /// </summary>
+        /// <param name="expected">Expected collection</param>
+        public AnyOfConstraint(IEnumerable expected)
+        {
+            _expected = expected;
+        }
+
+        /// <summary>
+        /// The Description of what this constraint tests, for
+        /// use in messages and in the ConstraintResult.
+        /// </summary>
+        public override string Description
+        {
+            get
+            {
+                return "any of " + MsgUtils.FormatValue(_expected);
+            }
+        }
+
+        /// <summary>
+        /// Test whether item is present in expected collection
+        /// </summary>
+        /// <typeparam name="TActual">Actual item type</typeparam>
+        /// <param name="actual">Actual item</param>
+        /// <returns>A <see cref="ConstraintResult"/> indicating whether item is present 
+        /// in expected collection.</returns>
+        public override ConstraintResult ApplyTo<TActual>(TActual actual)
+        {
+            var tolerance = Tolerance.Default;
+            foreach (var item in _expected)
+            {
+                if (_comparer.AreEqual(actual, item, ref tolerance))
+                {
+                    return new ConstraintResult(this, actual, ConstraintStatus.Success);
+                }
+            }
+
+            return new ConstraintResult(this, actual, ConstraintStatus.Failure);
+        }
+
+        #region Modifiers
+
+        /// <summary>
+        /// Flag the constraint to ignore case and return self.
+        /// </summary>
+        public AnyOfConstraint IgnoreCase
+        {
+            get
+            {
+                _comparer.IgnoreCase = true;
+                return this;
+            }
+        }
+
+        /// <summary>
+        /// Flag the constraint to use the supplied IComparer object.
+        /// </summary>
+        /// <param name="comparer">The IComparer object to use.</param>
+        /// <returns>Self.</returns>
+        public AnyOfConstraint Using(IComparer comparer)
```

Comment:
Do you have a use case in mind for "Using?" We shouldn't add it "just in case" it is needed because we have learned we can always add things but pretty much __never__ can remove them!

Reply:
It did cross my mind that we might want to consider having these equality-related constraints inherit a common base.

Reply:
Again, not sure if that is really needed for unit tests (same as .Using for .Equals, in fact). But might be useful for ui tests, For example, when example, when we add a lot of entries of 3 types, and that type is displayed with some formatting (uppercase/with index/etc), and formatting itself is not the point of the test.

Reply:
That makes me think of another use case for a collection-friendly `Is.AnyOf`: reusing values defined in a static readonly variable. Otherwise maintenance would be a pain.

Reply:
OK, you talked me into it! I'd still like an error for an empty array however.

Reply:
Great! Okay, let's do that error.